src/lazyarrays.jl

```diff
@@ -0,0 +1,67 @@
+using .LazyArrays: BroadcastArray, BroadcastVector, LazyArray
+
+# Necessary to make `BroadcastArray` work nicely with Zygote.
+function ChainRulesCore.rrule(config::ChainRulesCore.RuleConfig{>:ChainRulesCore.HasReverseMode}, ::Type{BroadcastArray}, f, args...)
+    return ChainRulesCore.rrule_via_ad(config, Broadcast.broadcasted, f, args...)
+end
```

Comment:
ChainRules definitions for LazyArrays.jl should really not be part of DistributionsAD. The other type piracy is already bad but at least Distributions-specific. But these lines seem really inappropriate here.

Reply:
Agreed! See https://github.com/JuliaArrays/LazyArrays.jl/issues/232

Think the plan is to make a glue-package for now.

Btw, this `rrule` is also not good since, for example, `logpdf(arraydist(BroadcastArray(Normal, x)), data)` will then be separated into two broadcast statements again, which is the opposite of what we want :confused: 

We _could_ of course define adjoint rules for `logpdf`, etc. that specializes on the `BroadcastArray` scenario, but that is all non-ideal :confused:

Reply:
I don't think a glue package is a good long term solution. If one does not want to load ChainRulesCore for all users (even though I think it's loaded anyways in almost all realistic scenarios), a weak dependency seems the best way IMO.

Reply:
Btw got any good idea of how to define the constructor for `BroadcastArray` properly?

Reply:
This far I have:
```julia
function ChainRulesCore.rrule(::Type{BroadcastArray}, f, args...)
    function BroadcastArray_pullback(Δ::ChainRulesCore.Tangent)
        return (ChainRulesCore.NoTangent(), Δ.f, Δ.args...)
    end
    return BroadcastArray(f, args...), BroadcastArray_pullback
end

function ChainRulesCore.rrule(
    config::ChainRulesCore.RuleConfig{>:ChainRulesCore.HasReverseMode},
    ::typeof(Distributions.logpdf),
    dist::Distributions.Product{V,D,A}, x::AbstractVector{<:Real}
) where {V,D<:Distribution,A<:BroadcastArray}
    cl = DistributionsAD.Closure(logpdf, DistributionsAD._inner_constructor(typeof(dist.v)))
    y, dy = ChainRulesCore.rrule_via_ad(config, broadcast, cl, x, dist.v.args...)
    z, dz = ChainRulesCore.rrule_via_ad(config, sum, y)

    f = dist.v.f
    function logpdf_adjoint(Δ...)
        # 1st argument is `sum` -> nothing.
        (_, sum_Δ...) = dz(Δ...)
        # 1st argument is `broadcast` -> nothing.
        # 2nd argument is `cl` -> `nothing`.
        # 3rd argument is `x` -> something.
        # Rest is `dist` arguments -> something
        (_, _, x_Δ, args_Δ...) = dy(sum_Δ...)
        # Construct the structural tangents.
        ba_tangent = ChainRulesCore.Tangent{A}(f=f, args=args_Δ)
        dist_tangent = ChainRulesCore.Tangent{typeof(dist)}(v=ba_tangent)

        return (ChainRulesCore.NoTangent(), dist_tangent, x_Δ)
    end

    return z, logpdf_adjoint
end
```
which does indeed to the trick but does not look great :confused: 

1. I don't like how I have to call back into AD, but I don't atm see a way around that.
2. I'm not too familiar with structural `Tangent`, and so I don't know if nesting them is a bad idea. For example I noticed that when we hit the pullback for `BroadcastArray`, I'm looking at a `Tangent{Any}` despite this not being the case in `logpdf_adjoint` (though this might just be Zygote doing something with it in the mean time?).
3. This isn't full support for `BroadcastArray` since in most cases it won't receive a `Tangent`.

Any more?

Reply:
And I should probably be using `ProjectTo` instead of explicit usage of `Tangent`?

Reply:
Awesome:) Btw, _really_ appreciate your responses over these past days; been incredibly helpful:) I'll be doing some workshop talks at a winter school on Turing next week and I would have loved to pin this down before that but we'll see :+1:

Reply:
Btw I just pushed the resulting impl so you can look at it properly when you have time. It _does_ seem to work awfully well but it's _very_ hacky :confused:

Reply:
A PhD student from Gothenburg actually told me about the winter school and thought it might be interesting for me based on my research interests. I would have liked to go to Norway since I have only been to Bergen so far but unfortunately I don't have time (and I assume I'm not the intended audience for a workshop about Turing :stuck_out_tongue:).